### PR TITLE
Allow X at the end of ISBN

### LINKED ISSUE #27659: ISBN validator should allow X as a last char
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

PrestaShop do not allow ISBN to have X as last char.
Such ISBN's are valid (example: 83-88945-22-X)
https://www.isbn-check.de/checkisbn.pl?isbn=83-88945-22-X&submit=test&lang=en
In definition: https://en.wikipedia.org/wiki/International_Standard_Book_Number for ISBN-10 X is for 10 for checksum control.

Probably classes->Validate.php->isIsbn regexp to be adjusted only.

### Expected behavior

ISBN should allow X as last char

### Steps to reproduce

1. Go to Products -> Add new Product
2. Go to options, add a valid ISBN with X at the end, e.g. 83-88945-22-X
3. PrestaShop will not allow to create product with such a ISBN

### PrestaShop version(s) where the bug happened

1.7.8.3

### PHP version(s) where the bug happened

_No response_

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by ghost**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | ISBN validator should allow X as a last char
| Type?             | bug fix
| Category?         |  CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/27659
| How to test?      | See beelow 
| Possible impacts? | The specs need to be update

**To test :** 

Correct ISBN values:

- ISBN 978-0-596-52068-7
- ISBN-13: 978-0-596-52068-7
- 978 0 596 52068 7
- 9780596520687
- ISBN-10 0-596-52068-9
- 0-596-52068-9
- 0-596-52068-X
- 978-0-596-52068-7

Incorrect ISBN value 

- 978-3-16-144100-X --> X is not valid for ISBN 13
- 0-8044-29X7-X --> X not at the end is not valid for ISBN 10

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27662)

### CODE REVIEW

File: classes/Validate.php [hunk 1]
```diff
@@ -682,7 +682,7 @@ public static function isEan13($ean13)
      */
     public static function isIsbn($isbn)
     {
-        return !$isbn || preg_match('/^[0-9-]{0,32}$/', $isbn);
+        return !$isbn || preg_match('/^[0-9][0-9-]{1,30}[0-9X]$/i', $isbn);
```
Comment (Progi1984) [inline classes/Validate.php]:
```suggestion
        return !$isbn || preg_match(Isbn::VALID_PATTERN, $isbn);
```

And define `use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\Isbn;`

**Review by Progi1984 — COMMENTED** [review]
And don't forget the unit test :)

File: src/Core/Domain/Product/ValueObject/Isbn.php [hunk 2]
```diff
@@ -38,7 +38,7 @@ class Isbn
     /**
      * Valid ISBN regex pattern
      */
-    public const VALID_PATTERN = '/^[0-9-]{0,32}$/';
+    public const VALID_PATTERN = '/^[0-9][0-9-]{1,30}[0-9X]$/i';
```
Comment (PierreRambaud) [inline src/Core/Domain/Product/ValueObject/Isbn.php]: This regex is not valid regarding the https://en.wikipedia.org/wiki/International_Standard_Book_Number
```suggestion
    public const VALID_PATTERN = '/^(?=(?:\D*\d){10}(?:(?:\D*\d){3})?X?$)[\d-X]+$/i';
```
Are you looking for something like that?

**Comment by ghost** [comment]
Thank you for reviews @Progi1984 && @PierreRambaud  i made changes

File: src/Core/Domain/Product/ValueObject/Isbn.php [hunk 3]
```diff
@@ -38,7 +38,9 @@ class Isbn
     /**
      * Valid ISBN regex pattern
      */
-    public const VALID_PATTERN = '/^[0-9-]{0,32}$/';
+    public const VALID_PATTERN = '/^(?:ISBN(?:-1[03])?:? )?(?=[0-9X]{10}
```
Comment (PierreRambaud) [inline src/Core/Domain/Product/ValueObject/Isbn.php]: If you're using https://www.oreilly.com/library/view/regular-expressions-cookbook/9781449327453/ch04s13.html you can maybe specify it 
or better, use the defined pattern with comments:
```
^
(?:ISBN(?:-1[03])?:?\ )?  # Optional ISBN/ISBN-10/ISBN-13 identifier.
(?=                       # Basic format pre-checks (lookahead):
  [0-9X]{10}$             #   Require 10 digits/Xs (no separators).
 |                        #  Or:
  (?=(?:[0-9]+[-\ ]){3})  #   Require 3 separators
  [-\ 0-9X]{13}$          #     out of 13 characters total.
 |                        #  Or:
  97[89][0-9]{10}$        #   978/979 plus 10 digits (13 total).
 |                        #  Or:
  (?=(?:[0-9]+[-\ ]){4})  #   Require 4 separators
  [-\ 0-9]{17}$           #     out of 17 characters total.
)                         # End format pre-checks.
(?:97[89][-\ ]?)?         # Optional ISBN-13 prefix.
[0-9]{1,5}[-\ ]?          # 1-5 digit group identifier.
[0-9]+[-\ ]?[0-9]+[-\ ]?  # Publisher and title identifiers.
[0-9X]                    # Check digit.
$
```

Comment (ghost) [reply src/Core/Domain/Product/ValueObject/Isbn.php]: I can't add comments or use a lot of line breaks the test fails; I added all that in the comment

Comment (PierreRambaud) [reply src/Core/Domain/Product/ValueObject/Isbn.php]: You can, by using concatenation

**Review by PierreRambaud — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by atomiix** [comment]
Changed waiting for QA to Waiting for PM as we're still waiting for the spec to be validated ;)

**Comment by MatShir** [comment]
The PR specs need to be fixed but the feature is validated 👍 . We can unleash the QA 🦁

**Comment by sLorenzini** [comment]
@PrestaShop/product-team hello 😀 ,
is this feature approved ?
Thanks

**Comment by NeOMakinG** [comment]
@okom3pom there is a conflict to be resolved

**Comment by MatShir** [comment]
> @PrestaShop/product-team hello 😀 , is this feature approved ? Thanks

Yes 👍

**Review by sLorenzini — DISMISSED** [review]
hello @okom3pom,
Tested on develop branch, 7.2.21 php version.

Tested different cases:

- 3 products : 

> - without letters 
> - 13 char 
> - 4 differents separators(978-0-596-52068-7/978 0 596 52068 7/9780596520687)

- 3 products : 

> - without letters
> - 10 char 
> - 3 separators (0-596-52068-9/0 596 52068 9/9780596520687)

- 3 products : 

> - with letters 
> - 13 char 
> - 4 differents separators(978-0-596-52068-X/978 0 596 52068 X/978059652068X)

- 3 products : 

> - without letters
> - 10 char 
> - 3 separators (0-596-52068-X/0 596 52068 X/978059652068X)

Product : 

- with case insensitive 13 and 10 char (0-596-52068-x / 978-0-596-52068-x)
- with different separators 13 and 10 char (0-596 52068x / 9780 596-52068-x)
- with other separators 13 and 10 char (0_596_52068_x / 9780_596_52068_x)
- with more or less char (0-596-58-x / 978-0-596-52111068-x)
- with other letter 13 and 10 char (0-596-52068-A / 978-0-596-52068-A)

QA approved ✅

**Comment by Progi1984** [comment]
@okom3pom Needs a rebase before mergind :)

**Review by Progi1984 — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by PululuK — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Pull request merged**